File: project/frontend/src/lib/nodes/definitions/pytorch/linear.ts
/**
 * PyTorch Linear Layer Node Definition
 */

import { NodeDefinition } from '../../base'
import { NodeMetadata, BackendFramework } from '../../contracts'
import { TensorShape, BlockConfig, ConfigField, BlockType } from '../../../types'

export class LinearNode extends NodeDefinition {
  readonly metadata: NodeMetadata = {
    type: 'linear',
    label: 'Linear',
    category: 'basic',
    color: 'var(--color-primary)',
    icon: 'Lightning',
    description: 'Fully connected layer',
    framework: BackendFramework.PyTorch
  }

  readonly configSchema: ConfigField[] = [
    {
      name: 'out_features',
      label: 'Output Features',
      type: 'number',
      required: true,
      min: 1,
      description: 'Number of output features'
    },
    {
      name: 'bias',
      label: 'Use Bias',
      type: 'boolean',
      default: true,
      description: 'Add learnable bias'
    }
  ]

  computeOutputShape(inputShape: TensorShape | undefined, config: BlockConfig): TensorShape | undefined {
    // Check if out_features is properly set (not undefined, null, or empty)
    const outFeatures = config.out_features
    if (!inputShape || outFeatures === undefined || outFeatures === null || outFeatures === '') {
      return undefined
    }

    if (inputShape.dims.length !== 2) {
      return undefined
    }

    const numOutFeatures = Number(outFeatures)
    if (isNaN(numOutFeatures) || numOutFeatures <= 0) {
      return undefined
    }

    return {
      dims: [inputShape.dims[0], numOutFeatures],
      description: 'Fully connected output'
    }
  }

  validateIncomingConnection(
    sourceNodeType: BlockType,
    sourceOutputShape: TensorShape | undefined,
    targetConfig: BlockConfig
  ): string | undefined {
    // Allow connections from input/dataloader without shape validation
    if (sourceNodeType === 'input' || sourceNodeType === 'dataloader') {
      return undefined
    }

    // Empty and custom nodes are flexible
    if (sourceNodeType === 'empty' || sourceNodeType === 'custom') {
      return undefined
    }

    // Validate dimension requirement
    return this.validateDimensions(sourceOutputShape, {
      dims: 2,
      description: '[batch, features]'
    })
  }
}

File: project/block_manager/services/nodes/frameworks/pytorch/nodes/linear.py
"""PyTorch Linear (Fully Connected) Layer Node"""

from typing import Any, Dict, List, Optional

from ....core.types import ConfigField, Framework, NodeMetadata, TensorShape
from ..base import PyTorchNodeDefinition


class LinearNode(PyTorchNodeDefinition):
    """
    Linear/Fully Connected Layer.

    Applies a linear transformation: y = xW^T + b

    Input: [batch_size, in_features]
    Output: [batch_size, out_features]
    """

    @property
    def metadata(self) -> NodeMetadata:
        return NodeMetadata(
            type="linear",
            label="Linear",
            category="basic",
            color="var(--color-primary)",
            icon="Lightning",
            description="Fully connected layer",
            framework=Framework.PYTORCH
        )

    @property
    def config_schema(self) -> List[ConfigField]:
        return [
            ConfigField(
                name="out_features",
                label="Output Features",
                type="number",
                required=True,
                min=1,
                description="Number of output features"
            ),
            ConfigField(
                name="bias",
                label="Use Bias",
                type="boolean",
                default=True,
                description="Add learnable bias"
            )
        ]

    def compute_output_shape(
        self,
        input_shape: Optional[TensorShape],
        config: Dict[str, Any]
    ) -> Optional[TensorShape]:
        # Check if out_features is properly set
        out_features = config.get("out_features")
        if not input_shape or out_features is None or out_features == '':
            return None

        if len(input_shape.dims) != 2:
            return None

        try:
            num_out_features = int(out_features)
            if num_out_features <= 0:
                return None
        except (ValueError, TypeError):
            return None

        return TensorShape(
            dims=[input_shape.dims[0], num_out_features],
            description="Fully connected output"
        )

    def validate_incoming_connection(
        self,
        source_node_type: str,
        source_output_shape: Optional[TensorShape],
        target_config: Dict[str, Any]
    ) -> Optional[str]:
        # Allow connections from input/dataloader without shape validation
        if source_node_type in ("input", "dataloader"):
            return None

        # Empty and custom nodes are flexible
        if source_node_type in ("empty", "custom"):
            return None

        # Validate dimension requirement
        return self.validate_dimensions(
            source_output_shape,
            2,
            "[batch, features]"
        )
